### ISSUE TITLE
Obscure error messages

### ISSUE THREAD

**Issue by NickBolles** [FusionAuth/fusionauth-typescript-client#8]
If the incorrect credentials are entered a 404 is returned, and the parsing of the response body fails. This results in an obscure error:
```
exception: FetchError: invalid json response body at https://auth.nickbolles.com/api/login reason: Unexpected end of JSON input
      at C:\Users\nickb\Resilio Sync\Projects\goalsapp\project\packages\server\node_modules\node-fetch\lib\index.js:272:32
      at processTicksAndRejections (internal/process/task_queues.js:93:5) {
    message: 'invalid json response body at https://auth.nickbolles.com/api/login reason: Unexpected end of JSON input',
    type: 'invalid-json'
  }
```

Seems like that should be handled by the client and throw a better error. Took me a bit to figure out that I just missed a character in my email address...  😑

**Comment by robotdan** [comment]
Thanks for the report @NickBolles - I want to say we've corrected this behavior. 

The `404` status code is correct, but when we receive that status code from FusionAuth, the client should expect the body length to be 0. 

If you still encounter this error let us know. Thanks!

**Comment by NickBolles** [comment]
I'll try it out. Thank you!

**Comment by snazzypants** [comment]
What version should be checking to see if this is fixed? I am also seeing this behavior - am on FusionAuth version 1.15.8 and @fusionauth/typescript-client 1.15.4

**Comment by robotdan** [comment]
@snazzypants you get a `404` and the client is attempting to deserialize the body and you see a similar exception as @NickBolles? 

Can you provide an example snippet of code and the response or error you're seeing? Thanks. 

@tyduptyler13 I was thinking this was something we fixed already. In the case of the Login API, or others - that may return a `404` w/out a body - the client should be able to handle that. Do you want to take a look when you get a free moment? Thanks.

**Comment by NickBolles** [comment]
@robotdan I haven't tried this again, but this seems like where the error would occur

```
/**
* A function that returns the JSON form of the response text.
*
* @param response
* @constructor
*/
async function ErrorJSONResponseHandler<ERT>(response: Response): Promise<ClientResponse<ERT>> {
  let clientResponse = new ClientResponse<ERT>();

  clientResponse.statusCode = response.status;
  clientResponse.exception = await response.json();

  return clientResponse;
}
```

Specifically this line:
```
clientResponse.exception = await response.json();
```

**Comment by snazzypants** [comment]
I've updated to 1.15.8 and still get this behavior.

Sample code using this library:
```
export function login(username: string, password: string, applicationId: string, tenantId: string) {
	client.setTenantId(tenantId);
	const requestBody = {
		loginId: username,
		password: password,
		applicationId
	}

	return client.login(requestBody)
	.then( clientResponse => {
		return {success response};
	})
	.catch( clientResponse => {
		console.log(clientResponse)
		return {error response}
	});
}
```

With invalid credentials, the console.log output in the catch block:
```
ClientResponse {
  statusCode: 404,
  exception: FetchError: invalid json response body at https://[url]/api/login reason: Unexpected end of JSON input
      at C:\jnl\app-backend\node_modules\node-fetch\lib\index.js:272:32
      at processTicksAndRejections (internal/process/task_queues.js:97:5) {     
    message: 'invalid json response body at https://[url]/api/login reason: Unexpected end of JSON input',
    type: 'invalid-json'
  }
}
```

When I hit the fusionauth server login endpoint directly with a wrong password, it returns a 404 with no body, as expected.

**Comment by robotdan** [comment]
@tyduptyler13 can you take a look and see if this is still an issue, or perhaps we have a second issue to resolve?

**Comment by tyduptyler13** [comment]
I refactored our typescript client to dynamically determine the response type. Previously we were statically compiling the response handler into every call but now it just generically always uses the json handler (which can skip json parsing if the body isn't returning the correct content type).

We just rolled a release so it might be a little bit of time before we publish the changes, however I wrote a test to explicitly trigger the issue you were having and it doesn't explode anymore. I'll let you close this whenever you get a chance to verify the changes.
